### DISCUSSION TITLE
Is it possible to direct data to the best BR using static IPv6 in a Thread network?

### DISCUSSION THREAD

**Discussion by BrendowAlisson** [openthread/openthread#9589] [Q&A] [answered]
Hi there,

I have a question about setting up 2 Border Routers (BRs) in a Thread network, and I'm struggling to find guidance on this. My main concern is ensuring that a device consistently sends data to the better-performing BR.

For example, I have a device using CoAP to send data to the IPv6 address 2001:dead:beef:0:0:0:0:1. If I add this address to both BRs, the network seems to work, and the device communicates with the BR it's connected to. However, when I add a router, things get tricky. The router sends an Address Query, which leads to duplicate IPv6 address detection for both BRs, as explained in Sections 5.6 and 5.10.1.1 of the Thread specification. This results in the removal of one BR's IPv6 address, as described in Section 5.6.4. 

Even if I re-add the IPv6 address, it doesn't work as intended because the router always sends data to the BR saved in its cache. Using multicast isn't an option because both BRs would receive the message, and I want the one with the better signal to get the data.

I was thinking of using fixed IPs destination CoAP messages for all MTD and letting the network choose the best BR. Is that possible? Your advice would be greatly appreciated.

**Comment by abtink** [comment, marked as answer]
@BrendowAlisson , devices on Thread mesh will choose the BR with lower path cost (effectively closer BR, fewer hops away) if there are multiple options when determining the destination (this is used when multiple BR provide same off-mesh route prefix). This is already part of spec and implementation.

> For example, I have a device using CoAP to send data to the IPv6 address 2001:dead:beef:0:0:0:0:1. If I add this address to both BRs, the network seems to work, and the device communicates with the BR it's connected to. However, when I add a router, things get tricky. The router sends an Address Query, which leads to duplicate IPv6 address detection for both BRs, as explained in Sections 5.6 and 5.10.1.1 of the Thread specification. This results in the removal of one BR's IPv6 address, as described in Section 5.6.4.

Adding the same unicast address on multiple BRs is not correct. It creates duplicate address error.

You may want to consider anycast address (which can be assigned to multiple devices and then sender would pick the closest destination to send to). Though they are not often used by application layer. You may want to read and learn how the "network data" service entries work and how the ALOC addresses are used for them.

**Comment by BrendowAlisson** [comment]
Hello @abtink 

I successfully implemented your suggestion, allowing me to use two BRs with the same CoAP address using DHCPv6 Agent. I assigned the IPv6 `fde5:8dba:82e1:1::ff:fe00:fc01` to both BRs and registered them in the network using the following commands:

```plaintext
> prefix add 2001:dead:beef:0::/64 prdos med
> netdata register
```

Everything is functioning well; the router is forwarding messages, and I'm receiving them. However, I have an additional question.

I assumed that once everything was set up, the data would be interleaved between BR1 and BR2 because the router would identify the "closest" DHCPv6 Agent, resulting in a variation or sending the message always to one BR. However, I've noticed that the router consistently forwards messages to the first BR registered in the network (in this case, `dc00`).

```plaintext
> netdata show
Prefixes:
2001:dead:beef:0::/64 prdos med dc00
2001:dead:beef:0::/64 prdos med c100
Routes:
Services:
Done
```
I observed that when `dc00` is considered the "closest" one, performing a `factoryreset` on `dc00` triggers `c100` to start sending messages. However, upon reconfiguring `dc00`, it does not resume sending messages, and I have yet to identify the reason behind this behavior.

Am I overlooking something, or is this the expected behavior? Your insights would be greatly appreciated.

**Reply by abtink** [reply]
FTD devices within Thread mesh will use their current view route table and path costs to select between multiple BR options (determine lowest path cost which often maps to closest in terms of hops away). So as long as the mesh remains stable and device's route table (path cost info) remains unchanged, the node will make the same decision.

**Reply by BrendowAlisson** [reply]
> So as long as the mesh remains stable and device's route table (path cost info) remains unchanged, the node will make the same decision.

I apologize for the confusion; I didn't understand this clearly.

Suppose there are two Border Routers (BRs), `BR 1` and `BR 2`. If `BR 1` is actively sending messages while `BR 2` is not, and you decide to perform a `factoryreset` on `BR 1`, resulting in `BR 2` taking over message transmission. When you reintroduce `BR 1` into the network afterward, will the FTD devices automatically revert to selecting `BR 1`? This was supposed to happen because there is a modification in the device's route table, right?

**Reply by abtink** [reply]
Yes. This is about FTD devices (on Thread mesh) and how they would pick if they see multiple options (e.g., multiple BRs providing the same external-route, ...).
